File: ob-zt/ac4a/nodejs/ac4a.js
'use strict';

const jwt = require('jsonwebtoken');

const ZT_AC4A = process.env.ZT_AC4A === 'true';
const JWT_SECRET = process.env.JWT_SECRET || 'dev-secret-change-me';

function _verify(call) {
  const vals = call.metadata.get('authorization');
  const auth = vals && vals[0];
  if (!auth || !auth.startsWith('Bearer ')) {
    const err = new Error('missing bearer token');
    err.code = 16; // UNAUTHENTICATED
    return err;
  }
  try {
    jwt.verify(auth.slice(7), JWT_SECRET, { algorithms: ['HS256'] });
    return null;
  } catch (e) {
    const err = new Error(e.message);
    err.code = 16;
    return err;
  }
}

function wrapService(impl) {
  if (!ZT_AC4A) return impl;
  const wrapped = {};
  for (const method of Object.keys(impl)) {
    const orig = impl[method];
    wrapped[method] = function (call, callback) {
      const err = _verify(call);
      if (err) {
        if (callback) return callback(err);
        call.destroy(err);
        return;
      }
      return orig.call(this, call, callback);
    };
  }
  return wrapped;
}

module.exports = { wrapService };

File: ob-zt/ac4a/python/ac4a.py
import os
import grpc
import jwt as pyjwt

ZT_AC4A = os.getenv("ZT_AC4A", "false").lower() == "true"
JWT_SECRET = os.getenv("JWT_SECRET", "dev-secret-change-me")
JWT_ALGORITHM = "HS256"


def _verify(context):
    meta = dict(context.invocation_metadata())
    auth = meta.get("authorization", "")
    if not auth.startswith("Bearer "):
        context.abort(grpc.StatusCode.UNAUTHENTICATED, "missing bearer token")
        return False
    try:
        pyjwt.decode(auth[7:], JWT_SECRET, algorithms=[JWT_ALGORITHM])
        return True
    except pyjwt.exceptions.InvalidTokenError as exc:
        context.abort(grpc.StatusCode.UNAUTHENTICATED, str(exc))
        return False


class Ac4aInterceptor(grpc.ServerInterceptor):
    def intercept_service(self, continuation, handler_call_details):
        if not ZT_AC4A:
            return continuation(handler_call_details)

        handler = continuation(handler_call_details)
        if handler is None:
            return None

        def _gate(fn):
            def wrapper(req_or_iter, context):
                if not _verify(context):
                    return None
                return fn(req_or_iter, context)
            return wrapper

        replacements = {
            field: _gate(getattr(handler, field))
            for field in ("unary_unary", "unary_stream", "stream_unary", "stream_stream")
            if getattr(handler, field) is not None
        }
        return handler._replace(**replacements)
